Switch PG::Work to Future in default_comm_hooks.cpp

agolynski opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59398 Switch PG::Work to Future in default_comm_hooks.cpp**

Differential Revision: [D28876182](https://our.internmc.facebook.com/intern/diff/D28876182)

agolynski committed:
Switch PG::Work to Future in default_comm_hooks.cpp

[ghstack-poisoned]

agolynski commented:
@agolynski has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28876182).

wayi1 reviewed (approved):
Thanks for the fix!

SciPioneer commented on torch/lib/c10d/default_comm_hooks.cpp:
```diff
@@ -9,39 +9,42 @@ namespace c10d {
-  auto allreduce_work = state_->allreduce(tensors);
+  auto allreduce_fut = state_->allreduce(tensors)->getFuture();
 
   // FIXME Access the result through the Future passed as argument, instead of
```
Please remove the FIXME comment.

agolynski replied:
Done

lw commented on torch/lib/c10d/default_comm_hooks.cpp:
```diff
@@ -9,39 +9,42 @@ namespace c10d {
                                     this](c10::ivalue::Future& /* unused */) {
-    auto tensor = allreduce_work->result()[0] / state_->getSize();
+
+    auto result = allreduce_fut->value();
+    TORCH_INTERNAL_ASSERT(result.isTensorList(),
+        "ProcessGroup::allreduce should return TensorList");
+    auto tensor = result.toTensorVector()[0] / state_->getSize();
     return c10::IValue(tensor);
   };
 
-  auto fut = allreduce_work->getFuture();
-  return fut->then(div_by_process_group_size, fut->elementType());
+  return allreduce_fut->then(div_by_process_group_size, allreduce_fut->elementType());
```
This code introduces a (potential) memory leak: by adding a callback, you end up storing on `allreduce_fut` a lambda whose closure contains an owning pointer to `allreduce_fut` itself. This creates a reference cycle, which means that if for some reason the future is never completed it will never be "garbage collected" because its refcount will never reach 0.

This is precisely the problem that was being addressed in the FIXME that you removed: instead of capturing `allreduce_fut`, we should use the argument that is being passed to the lambda, which also points to `allreduce_fut`, but which doesn't cause the reference cycle.

In other words, please do this:
```
  auto div_by_process_group_size = [this](c10::ivalue::Future& allreduce_fut) {
    auto result = allreduce_fut.value();
    ...
  };
  return allreduce_fut->then(div_by_process_group_size, ...);
```

The same applies to the other FIXME just below.

lw replied:
To go the extra mile, it might also be good to avoid capturing `this` in the lambda, because it boils down to capturing a raw pointer, which means that there is no guarantee that the pointed-to object will still be alive once the callback fires.

The only reason for capturing `this` is to access `state_->getSize()` (I believe), hence what we could do instead is capturing that directly by value:
```
auto div_by_process_group_size = [size{state_->getSize()}](c10::ivalue::Future& allreduce_fut) { ... };
```

agolynski replied:
> The only reason for capturing `this` is to access `state_->getSize()` (I believe), hence what we could do instead is capturing that directly by value:

thanks, capturing 'this' here is not a good idea indeed, this can easily lead to a crash when Future is not immediately returned (e.g. GLOO and MPI backends)

agolynski replied:
> This code introduces a (potential) memory leak: by adding a callback, you end up storing on `allreduce_fut` a lambda whose closure contains an owning pointer to `allreduce_fut` itself. This creates a reference cycle, which means that if for some reason the future is never completed it will never be "garbage collected" because its refcount will never reach 0.
> 
> This is precisely the problem that was being addressed in the FIXME that you removed: instead of capturing `allreduce_fut`, we should use the argument that is being passed to the lambda, which also points to `allreduce_fut`, but which doesn't cause the reference cycle.
> 
> In other words, please do this:
> 
> ```
>   auto div_by_process_group_size = [this](c10::ivalue::Future& allreduce_fut) {
>     auto result = allreduce_fut.value();
>     ...
>   };
>   return allreduce_fut->then(div_by_process_group_size, ...);
> ```
> 
> The same applies to the other FIXME just below.

Done.

Talked offline: this is a defensive measure as if Future is not returned, typically behavior is not defined if we depend on result of that future. Memory leak will be the least of problems in this case.

agolynski committed:
Update on "Switch PG::Work to Future in default_comm_hooks.cpp"


Differential Revision: [D28876182](https://our.internmc.facebook.com/intern/diff/D28876182)

[ghstack-poisoned]

agolynski committed:
Update on "Switch PG::Work to Future in default_comm_hooks.cpp"


Differential Revision: [D28876182](https://our.internmc.facebook.com/intern/diff/D28876182)

[ghstack-poisoned]

agolynski committed:
Update on "Switch PG::Work to Future in default_comm_hooks.cpp"


Differential Revision: [D28876182](https://our.internmc.facebook.com/intern/diff/D28876182)

[ghstack-poisoned]

agolynski commented:
@agolynski has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28876182).

lw reviewed (approved):
Looks good, thanks!

agolynski committed:
Update on "Switch PG::Work to Future in default_comm_hooks.cpp"


Differential Revision: [D28876182](https://our.internmc.facebook.com/intern/diff/D28876182)

[ghstack-poisoned]

agolynski commented:
@agolynski has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28876182).
